src/common/k8s-api/kube-watch-api.ts

```diff
@@ -25,156 +25,185 @@
-          this.log({
-            message: new Error("Loading stores has failed"),
-            meta: { stores, error, options: opts },
+    let childController = new WrappedAbortController(parent);
+    const unsubscribe = disposer();
+
+    const loadThenSubscribe = async (namespaces: string[]) => {
+      try {
+        await store.loadAll({ namespaces, reqInit: { signal: childController.signal }, onLoadFailure });
+        unsubscribe.push(store.subscribe({ onLoadFailure, abortController: childController }));
+      } catch (error) {
+        if (!(error instanceof DOMException)) {
+          this.log(Object.assign(new Error("Loading stores has failed"), { cause: error }), {
+            meta: { store, namespaces },
           });
-        });
-      } else {
-        subscribe();
+        }
       }
+    };
 
-      // reload stores only for context namespaces change
-      cancelReloading = reaction(() => this.context?.contextNamespaces, namespaces => {
-        preloading?.cancelLoading();
-        unsubscribeStores();
-        preloading = load(namespaces);
-        preloading.loading.then(subscribe);
-      }, {
-        equals: comparer.shallow,
-      });
-    }
+    loadThenSubscribe(namespaces);
+
+    const cancelReloading = watchChanges
+      ? reaction(
+        // Note: must slice because reaction won't fire if it isn't there
+        () => this.context.contextNamespaces.slice(),
+        namespaces => {
+          console.log(`changing watch ${store.api.apiBase}`, namespaces);
+          childController.abort();
+          unsubscribe();
+          childController = new WrappedAbortController(parent);
+          loadThenSubscribe(namespaces);
+        },
+        {
+          equals: comparer.shallow,
+        },
+      )
+      : noop; // don't watch namespaces if namespaces were provided
 
-    // unsubscribe
     return () => {
-      if (isUnsubscribed) return;
-      isUnsubscribed = true;
-      cancelReloading();
-      preloading?.cancelLoading();
-      unsubscribeStores();
+      if (this.#watch.dec(store) === 0) {
+        // only stop the subcribe if this is the last one
+        cancelReloading();
+        childController.abort();
+        unsubscribe();
+      }
     };
   }
 
-  protected log({ message, cssStyle = "", meta = {}}: IKubeWatchLog) {
-    if (isProduction && !isDebugging) {
+  subscribeStores(stores: KubeObjectStore<KubeObject>[], { namespaces, onLoadFailure }: KubeWatchSubscribeStoreOptions = {}): Disposer {
+    const parent = new AbortController();
+    const unsubscribe = disposer(
+      ...stores.map(store => this.subscribeStore({
+        store,
+        parent,
+        watchChanges: !namespaces,
+        namespaces: namespaces ?? this.context?.contextNamespaces ?? [],
+        onLoadFailure,
+      })),
+    );
+
+    // unsubscribe
+    return once(() => {
+      parent.abort();
+      unsubscribe();
+    });
+  }
+
+  protected log(message: any, meta: any) {
+    if (isProduction) {
```

Comment:
why don't we log if `isProduction===true`? was it like this previously too?

Reply:
Good catch, I did subtly (and accidentally) change this. I think we probably should just remove this, and I will.